eth-rpc/serverdispatch.nim

```diff
@@ -0,0 +1,295 @@
+  let res = jsonValid(line, node)
+  if not res[0]:
+    return some((rjeInvalidJson, res[1]))
+  if not node.hasKey("jsonrpc"):
+    return some((rjeVersionError, ""))
+  if not node.hasKey("method"):
+    return some((rjeNoMethod, ""))
+  if not node.hasKey("id"):
+    return some((rjeNoId, ""))
+  return none(RpcJsonErrorContainer)
+
+proc wrapReply*(id: JsonNode, value: JsonNode, error: JsonNode): JsonNode =
+  return %{"jsonrpc": %"2.0", "result": value, "error": error, "id": id}
+
+proc sendError*(client: StreamTransport, code: int, msg: string,
+                id: JsonNode, data: JsonNode = newJNull()) {.async.} =
+  ## Send error message to client
+  let error = %{"code": %(code), "message": %msg, "data": data}
+  ifDebug: echo "Send error json: ",
+                wrapReply(newJNull(), error, id).pretty & "\c\l"
+  var answer = $wrapReply(id, newJNull(), error)
+  answer.add("\c\l")
+  result = client.write(addr answer[0], len(answer))
+
+proc sendJsonError*(state: RpcJsonError, client: StreamTransport,
+                    id: JsonNode, data = newJNull()) {.async.} =
+  ## Send client response for invalid json state
+  let errMsgs = jsonErrorMessages[state]
+  await client.sendError(errMsgs[0], errMsgs[1], id, data)
+
+proc processMessage(server: RpcServer, client: StreamTransport,
+                    line: string) {.async.} =
+  var node, id: JsonNode
+  # set up node and/or flag errors
+  var jsonErrorState = checkJsonErrors(line, node)
+
+  if jsonErrorState.isSome:
+    let errState = jsonErrorState.get
+    if errState.err == rjeInvalidJson:
+      # id cannot be retrieved
+      id = newJNull()
+    else:
+      id = if "id" in node: node["id"] else: newJNull()
+    await errState.err.sendJsonError(client, id, %errState.msg)
+  else:
+    let methodName = node["method"].str
+    id = node["id"]
+
+    if not server.procs.hasKey(methodName):
+      await client.sendError(METHOD_NOT_FOUND, "Method not found", id,
+                             %(methodName & " is not a registered method."))
+    else:
+      let res = await server.procs[methodName](node["params"])
+      var answer = $wrapReply(id, res, newJNull())
+      answer.add("\c\l")
+      discard await client.write(cast[pointer](addr answer[0]), len(answer))
+
+proc processClient(server: StreamServer,
+                   client: StreamTransport, udata: pointer) {.async.} =
+
+  while true:
+    let line = await client.readLine(limit = 1024)
+    if len(line) == 0:
+      client.close()
+      break
+
+    ifDebug: echo "Process client: ", $client.remoteAddress, ":" & line
+    var rpcsrv = cast[RpcServer](udata)
+    var fut = processMessage(rpcsrv, client, line)
+    yield fut
+    if fut.failed:
+      let err = fut.readError
+      if err of RpcProcError:
+        let rpcerr = RpcProcError(err)
+        await client.sendError(rpcerr.code, rpcerr.msg, rpcerr.data)
+      else:
+        await client.sendError(SERVER_ERROR, "Error", %getCurrentExceptionMsg())
```

Comment:
Shouldn't this still extract the message from the `err` value instead of calling `getCurrentExceptionMsg`? I don't think calling `getCurrentExceptionMsg` outside of an `except` handler is a well-defined operation.